### DISCUSSION TITLE
Specifying relations to load in QueryBuilder

### DISCUSSION THREAD

**Discussion by blitss** [mikro-orm/mikro-orm#5394] [Q&A]
I'm building a library that would allow to transform GraphQL AST to the mikro orm query (solving N+1 problems, including only necessary fields in the query and joining required relations). 

So let's imagine I have entity Book, which has nested relations "author" -> "company" and embeddable "metadata". I have a resulting query from AST parser which would indicate what fields/relations need to be loaded, which goes below.

```typescript
Book<{
name: true,
metadata: {
  createdAt: true,
  updatedAt: true
},
author: {
  id: true,
  userName: true,
  company: { 
    id: true,
    name: true,
  }
}
}>
```

As you can see, fields which are marked as true are the ones that need to be loaded and relations/embeddables are marked as object with the same list of fields. 

So what I want from MikroORM query builder is to return an entity of the same shape as this.

## The problem

1) In typeorm in queryBuilder you can call .setFindOptions and just specify relations you need and fields you need to select. It doesn't look like an option in MikroORM, there's something called `QueryBuilder.populate` but I think it serves a different purpose rather than to include relations.

Is using `leftJoin` for each relation the only viable option, or is there a better alternative? In that case I also have to use `EntityMetadata` to distinct embeddables because I don't need to join them. For embeddable looks like I can simply call `.select([{ field: '<embeddable name>.<embeddable field>' }])` and it seems work fine, can't do the same for relations.

2) How can I map path in `orderBy`, like `author.company.name` or `metadata.createdAt`, especially if it's nested? Do I have to go through EntityMetadata and understand the field name on my own (make separate mapper) or is there a better solution ?

**Comment by B4nan** [comment]
> In typeorm in queryBuilder you can call .setFindOptions and just specify relations you need and fields you need to select. It doesn't look like an option in MikroORM

Why do you even want to use a QB here? There are no FindOptions for QB, as there is no find method. You need to be explicit if you want to use QB, so instead of `populate` hints you call `joinAndSelect` (where you can also specify which fields to load).

> there's something called QueryBuilder.populate but I think it serves a different purpose rather than to include relations.

That's rather internal, not intended for public usage.

I would use `em.find` here, no need for QB. This is not typeorm, you don't need QB most of the time - in fact, there are very few things that would require QB. Based on what you are asking here , it feels like you don't really want a QB anyway.

> How can I map path in orderBy, like author.company.name or metadata.createdAt, especially if it's nested? Do I have to go through EntityMetadata and understand the field name on my own (make separate mapper) or is there a better solution ?

I am not sure what you are asking here? `qb.orderBy()` supports nesting (it supports the same input as `em.find`), and you pass there property names, no need to deal with field names translation yourself (anywhere in the ORM, including all QB methods). 

---

Anyway, two existing libraries to check, as it feels like you want to build something similar/connected:
- https://github.com/driescroons/graphql-fields-to-relations 
- https://github.com/darkbasic/mikro-orm-dataloaders

**Reply by blitss** [reply]
Wow, thanks for such quick reply on a weekend!

Yep the libraries are similar to what I need but my use case is more wide than they have to offer. You're correct that there's no need for QB here, but that's just one layer over which other stuff is implemented using some raw sql fragments. I just thought QueryBuilder would provide more flexibility for this rather than non-QB solution and it appeared as a cleaner option to me at first.

> I am not sure what you are asking here? qb.orderBy() supports nesting (it supports the same input as em.find), and you pass there property names, no need to deal with field names translation yourself (anywhere in the ORM, including all QB methods).

You're correct, I was trying to use path as a string, instead I have to use it as an object and then it works fine.

Anyways, thanks for the info and I'll try to do some research if I really need QB or maybe I can use non-QB solution. I'd like to keep this discussion open for a bit so I can follow up later if that is okay by you

**Reply by B4nan** [reply]
Sure thing, will be happy to assist with this, as well as promote the final package in the ORM docs once it's ready.
